File: ios/Accentury/Upload/UploadStatusBar.swift
import AccenturyCore
import SwiftUI

/// 업로드 상태 바 (KAN-100, KAN-147). 안드로이드 `upload/UploadStatusBar.kt`의 이식본이다.
///
/// 성공(Done)은 조용히 넘어가고, 진행 중 개수와 실패 건의 복구 경로만 보여준다. 요약 판정은
/// Core ``AccenturyCore/summarize(_:)``가 하고 여기는 그리기만 한다 — 안드로이드는 그 함수가
/// 같은 파일에 있지만 화면 요소가 아니라 갈림이라 이쪽에서는 화면 밖 순수 계층에 있다.
///
/// **복구 경로는 [재시도] 하나다** (KAN-147). [테스트 종료] 버튼은 없다 — 이탈 UX는 KAN-39
/// 디자인 때 정한다. 재시도 불가 실패에는 버튼을 주지 않는다: 같은 바이트를 다시 보내도 결과가
/// 같으므로 버튼이 거짓말이 된다. 그 대신 서버가 준 문구가 행에 그대로 남는다.
///
/// 오버레이가 아니라 WebView **아래**에 놓인다 — 녹음 중에도 실패한 업로드의 재시도 통로가
/// 가려지지 않아야 한다. 보여줄 것이 없으면 자리 자체를 차지하지 않는다.
struct UploadStatusBar: View {

    let entries: [UploadEntry]
    let labelOf: (String) -> String
    let onRetry: (String) -> Void

    var body: some View {
        let summary = summarize(entries)
        if !summary.isEmpty {
            VStack(alignment: .leading, spacing: Papercut.space1) {
                if summary.inFlight > 0 {
                    Text("업로드 중 \(summary.inFlight)건")
                        .papercutType(.label)
                        .foregroundColor(Papercut.ink)
                }

                ForEach(summary.failed, id: \.attemptId) { row in
                    HStack(spacing: Papercut.space2) {
                        Text("\(labelOf(row.attemptId)) 업로드 실패 — \(row.failure.message ?? "알 수 없는 오류")")
                            .papercutType(.caption)
                            .foregroundColor(Papercut.ink)
                            .frame(maxWidth: .infinity, alignment: .leading)
                        if row.failure.retryable {
                            AccenturyButton(text: "재시도", variant: .secondary) { onRetry(row.attemptId) }
                        }
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, Papercut.space4)
            .padding(.vertical, Papercut.space2)
            .background(Papercut.cream)
            // 위 화면(WebView·녹음 오버레이)과 이 바를 가르는 줄. 상자 테두리보다 얇고 흐리다.
            .overlay(alignment: .top) {
                Papercut.muted.frame(height: Papercut.borderHairline)
            }
        }
    }
}

File: app/src/main/java/com/accentury/app/upload/UploadStatusBar.kt
package com.accentury.app.upload

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.accentury.app.ui.components.AccenturyButton
import com.accentury.app.ui.components.ButtonVariant
import com.accentury.app.ui.theme.Spacing

/**
 * 성공(Done)은 조용히 넘어가고, 진행 중 개수와 실패 건의 복구 경로만 보여준다.
 *
 * 복구 경로는 [재시도] 하나다. 재시도 불가 실패에도 출구를 두지 않는다 (KAN-191, ux-ui.md
 * §4-D·§4-F): 이 바는 **정상 진행 중에 떠 있는 화면**이라 여기에 이탈 버튼을 달면 KAN-147이
 * 지운 것을 되살리는 셈이고, 업로드 한 건 때문에 아직 멀쩡한 응시 전체에 나가는 문이 생긴다.
 * 흐름은 여기서 끝나지 않는다 — 남은 문항을 마저 풀면 분석 대기 화면으로 이어지고 출구는
 * 거기 있다([다시 녹음] / 세션 단위 실패면 [다시 테스트하기]).
 *
 * 무디자인 하네스다 — 여기서는 상태가 보이는지만 확인한다.
 */
@Composable
fun UploadStatusBar(
    uploads: Map<String, UploadState>,
    labelOf: (attemptId: String) -> String,
    onRetry: (attemptId: String) -> Unit,
    modifier: Modifier = Modifier,
) {
    val summary = summarize(uploads)
    if (summary.inFlight == 0 && summary.failed.isEmpty()) return

    Column(
        modifier = modifier.fillMaxWidth().padding(horizontal = Spacing.x4, vertical = Spacing.x2),
        verticalArrangement = Arrangement.spacedBy(Spacing.x1),
    ) {
        if (summary.inFlight > 0) {
            Text("업로드 중 ${summary.inFlight}건", style = MaterialTheme.typography.labelLarge)
        }

        summary.failed.forEach { (attemptId, failed) ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Spacing.x2),
            ) {
                Text(
                    "${labelOf(attemptId)} 업로드 실패 — ${failed.message ?: "알 수 없는 오류"}",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onErrorContainer,
                    modifier = Modifier.weight(1f),
                )
                // 재시도 불가 실패는 같은 바이트를 다시 보내도 결과가 같으므로 버튼을 주지 않는다.
                if (failed.retryable) {
                    AccenturyButton(
                        text = "재시도",
                        variant = ButtonVariant.Secondary,
                        onClick = { onRetry(attemptId) },
                    )
                }
            }
        }
    }
}

internal data class UploadSummary(
    val inFlight: Int,
    val failed: List<Pair<String, UploadState.Failed>>,
)

/** 표시 로직의 순수한 부분. 실패 목록은 업로드를 넣은 순서를 그대로 따른다. */
internal fun summarize(uploads: Map<String, UploadState>): UploadSummary = UploadSummary(
    inFlight = uploads.values.count { it is UploadState.InFlight },
    failed = uploads.entries.mapNotNull { (attemptId, state) ->
        (state as? UploadState.Failed)?.let { attemptId to it }
    },
)
